partiql-planner/src/main/kotlin/org/partiql/planner/catalog/Function.kt

```diff
@@ -0,0 +1,115 @@
+package org.partiql.planner.catalog
+
+import org.partiql.planner.internal.SqlTypes
+import org.partiql.types.PType
+
+/**
+ * A [Function] is a PartiQL-routine callable from an expression context.
+ */
+public sealed interface Function {
+
+    /**
+     * The routine name. Required.
+     */
+    public fun getName(): String
+
+    /**
+     * The formal argument definitions. Optional.
+     */
+    public fun getParameters(): Array<Parameter> = DEFAULT_PARAMETERS
+
+    /**
+     * The function return type. Required.
+     */
+    public fun getReturnType(): PType.Kind
+
+    /**
+     * Compute a [PType] from the given arguments.
+     */
+    public fun computeReturnType(args: List<PType>): PType = SqlTypes.from(getReturnType())
```

Comment:
I've been thinking about this a bit, and I'm curious what you've been thinking about the following situation:
```
CREATE FUNCTION FOO(arg1 DECIMAL(15, 5)) RETURN DECIMAL(15, 5);
CREATE FUNCTION FOO(arg1 DECIMAL(10, 5)) RETURN DECIMAL(10, 5);
CREATE FUNCTION FOO(arg1 DECIMAL(5, 3)) RETURN DECIMAL(5, 3);

FOO(1); -- where the literal 1 is an INTEGER.
```

I imagine that, in the above function definitions, this would be represented by a single function signature (described with parameters of PType.Kind and a return type of PType.Kind). Now, my question is: when we have arguments that are coercible to a function's parameters' kinds, which specific variant of a kind will we choose? In this scenario, what would INT get coerced to?

Reply:
It would be coerced to the PType which is created by `SqlTypes.fromKind(kind)`.

For you specific example, it would be an arbitrary decimal.

`PType.fromKind(Kind.DECIMAL) -> PType.decimalArbitrary()`

Reply:
Interesting -- so coercions have some pre-determined mapping. So, in this model, now that we've coerced the literal integer to a decimal arbitrary, how do we know what the output type will be?
```
FOO(1::decimal_arbitrary) -> ?
```

I'm just mentally trying to understand how this works for coercions, and since I've recently done work on some of the arithmetic operators, I'm curious to see if this modeling is flexible enough to implement the operators with this logic.